proposals/0000-namespace-specified-imports.rst

```diff
@@ -0,0 +1,554 @@
+module identifier.
+
+More concretely, in the grammar accepted by GHC, ::
+
+      importdecl -> 'import' [src] ['safe'] ['qualified'] [package] modid ['qualified'] ['as' modid] [impspec]
+
+is changed to ::
+
+      importdecl -> 'import' [src] ['safe'] ['qualified'] [package] modid [namespace] ['qualified'] ['as' modid] [impspec]
+
+      namespace  -> 'data'
+                 |  'type'
+
+With a namespace specified in the import, only identifiers belonging to the
+corresponding namespace will be brought into the scope, as if an explicit import
+list was given mentioning only those identifiers (with the namespace specifier
+on each item).
+
+If an import declaration uses both a namespace specifier and an explicit import
+list, the explicit import list may not mention a different namespace specifier,
+nor an identifier that is not available in the given namespace, otherwise a name
+resolution error will be reported.  It is allowed to redundantly specify the
+same namespace specifier on the import declaration and on an individual item.
+
+If an import declaration uses a namespace specifier but no explicit import list,
+it is not an error for the declaration to bring no names into scope,
+e.g. because the ``data`` specifier was used on a module that exports only type
+names. (GHC may of course warn that such an import is redundant.)
+
+
+Use ``data`` specifier in fixity declarations and ``WARNING``/``DEPRECATED`` pragmas
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This proposal changes `GHC Proposal #65
+<https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0065-type-infix.rst>`__
+to use the ``data`` namespace specifier instead of ``value``.  (The specific
+changes are thanks to Vladislav Zavialov and form part of the PR.)
+
+That proposal has not yet been implemented, so this is not a breaking change.
+
+
+Examples
+---------
+
+
+Import/export lists with namespace specifiers
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. code:: haskell
+
+   {-# LANGUAGE ExplicitNamespaces, TypeFamilies #-}
+   {-# OPTIONS_GHC -Wpattern-namespace-specifier #-}
+   module M
+     ( D            -- Accepted: exports data family D
+     , data D       -- Accepted: exports data constructor D
+     , C(type D)    -- Accepted: exports class C and data family D
+     , D(data f)    -- Accepted: exports data family D and field f
+     , pattern D    -- Accepted: exports data constructor D but emits warning
+     , T(data D)    -- Accepted: exports type T and data constructor D
+     , data f       -- Accepted: exports field f
+     , data v       -- Accepted: exports term v
+     , type T (..)  -- Accepted: exports type T and all its constructors
+     , T(pattern D) -- Rejected: pattern keyword cannot be used in sub-list
+     , data T       -- Rejected: T not in scope in data namespace
+     , type E       -- Rejected: E not in scope in type namespace
+     ) where
```

Comment:
Can we also have a `TypeData` example here? I assume that if I have `type data K = T1`, then I must export/import it as `K(type T1)` rather than `K(data T1)`.

Reply:
Good idea! Yes my reading of #106 is that the names are introduced in the type namespace, so would use `type` rather than `data`.

Reply:
I added `TypeData` to Effects and Interactions (https://github.com/adamgundry/ghc-proposals/blob/namespace-specified-imports/proposals/0000-namespace-specified-imports.rst#41typedata).